### ISSUE TITLE
It should be possible to write a short script that uses body with gopherjs serve.

### ISSUE THREAD

**Issue by shurcooL** [gopherjs/gopherjs#571]
Currently, I don't know of a way to write a short script for `gopherjs serve` that accesses the body, because the script can execute before the body is available.

It might be because I can't think of a way to do it, or because the default HTML served by `gopherjs serve` is not ideal.

We should find a good way. I'd like to be able to show beginners how easy it is to get started with `gopherjs serve` and writing a short script, and not have to get into the subtle details of order execution until more advanced lessons.

I tried to resolve this via #570, where I wrote:

> it might still be better for `gopherjs serve` to optimize for beginners and short snippets, rather than performance and advanced features (like avoiding a flash of empty body).

**Comment by flimzy** [comment]
Are the standard onload solutions used by js/jquery  insufficient?

**Comment by shurcooL** [comment]
I don't want to have to rely on jQuery. Without it, this is the shortest solution that works in all cases that I know of:

```Go
var document = dom.GetWindow().Document().(dom.HTMLDocument)

func main() {
	switch readyState := document.ReadyState(); readyState {
	case "loading":
		document.AddEventListener("DOMContentLoaded", false, func(dom.Event) {
			go setup()
		})
	case "interactive", "complete":
		setup()
	default:
		panic(fmt.Errorf("internal error: unexpected document.ReadyState value: %v", readyState))
	}
}

func setup() {
	body := document.Body()
	// ... use body right away
}
```

@neelance made a suggestion in https://github.com/gopherjs/gopherjs/pull/570#issuecomment-271715842:

> Maybe try to add this before the other script tag:
`<script>document.body = document.createElement("body");</script>`

It sounds interesting, I'll try/consider it.

**Comment by slimsag** [comment]
I read @flimzy 's message as asking what is wrong with the standard onload solution jquery uses -- not suggesting jquery itself should actually be used.

One idea would be to pack the snippet you pasted above into a library:

```
var document = dom.GetWindow().Document().(dom.HTMLDocument)

func main() {
	dom.Ready(func() {
		body := document.Body()
		// ... use body
	})
}
```

**Comment by flimzy** [comment]
> I don't want to have to rely on jQuery. Without it, this is the shortest solution that works in all cases that I know of:

No, I wouldn't suggest depending on jQuery, either... just using the same approach, with a snippet like the one you pasted.  But as @slimsag suggested, I'd stick it in a library somewhere. (I do exactly that for my own GopherJS projects).

And as common as the need is, and as standard as the solution is, putting it in a library makes sense to me.

**Comment by shurcooL** [comment]
I see, that makes more sense.

I don't know of `dom` is the right package for this because it doesn't offer utilities, just straight mapping to DOM (/cc @dominikh), but I understand the idea.

If not `dom`, what would be a good place?

**Comment by flimzy** [comment]
I wonder if this might be something to live in its own, tiny package. It's a pretty common need, regardless of any other frameworks. It could be nice not to force someone to depend on the `dom` package (or any other larger one) if this is the only thing they need.

I wrote such a module that I use which you can see [here](https://github.com/flimzy/onload).  (Not guaranteed to cover all use cases, and not really intended for public consumption--just something for me to use.)

**Comment by alltom** [comment]
The body tag would be accessible to the JavaScript immediately if the script tag were inside a body tag. Does `gopherjs serve` send a body tag before the script tag?

```js
<body>
<script>alert(document.body);</script>
```

**Comment by shurcooL** [comment]
That's an interesting idea.

But I think I see a pitfall with that approach. Doing something like `js.Global.Get("document").Get("body").Set("innerHTML", renderedBodyHTML)` would cause the script to get removed from DOM, wouldn't it?

**Comment by alltom** [comment]
Yeah, but it's already loaded and executed by then. I don't think it
affects a script for its tag to be removed from the DOM.

**Comment by shurcooL** [comment]
I thought it did, but I could be wrong. I'll need to check it again. Thanks for the suggestion.
